### PR TITLE
[WIP]: Use demo.py to implement tutorials

### PR DESCRIPTION
**Pull request by shoeb-github**
xref: #696 

This is an implementation of `tutor.py` using  `demo.py`.

The approach is to extend the `DemoPath` class to also parse file extensions other than `.py` like `.rst`, `.htm`. However, this does not look clean. It is probably better to do this unification as outlined in issue  [#24](https://github.com/enthought/ets-examples/issues/24) of ETS examples.

Things that don't work:
- no support for `.desc`, `.swf` files.

To do:
- [ ] Add `next`, `previous` buttons in the UI to navigate the tutorial directory structure.

### CODE REVIEW

**Comment by corranwebster** [comment]
Rather than subclassing, how much of this functionality can be merged into the `demo.py` classes? In other words, can we enhance the demos with tutor functionality without breaking them?

Or alternatively, how much functionality can we drop from the tutor.py code?

Also, I don't think we have _any_ actual tutorials using videos or similar, so we can drop support for a lot of the file formats without losing anything.

File: traitsui/extras/tutor.py [hunk 1]
```diff
@@ -873,729 +191,285 @@ def run_code(self):
...
 
-    def _add_htm_item(self, path):
-        """ Creates a description item for an HTML file.
+    def _parse_html_file(self, path, name):
+        """ Parse a html file
         """
-        # Check if there is a corresponding .rst (restructured text) file:
-        dir, base_name = os.path.split(path)
-        rst = os.path.join(dir, os.path.splitext(base_name)[0] + '.rst')
+        return self._parse_htm_file(path, name)
```
Comment (corranwebster) [inline traitsui/extras/tutor.py]: This just does infinite recursion? (it looks like the original version did as well)

I think we can just remove this.

File: traitsui/extras/tutor.py [hunk 2]
```diff
@@ -873,729 +191,285 @@ def run_code(self):
...
-    # Should the Python code be automatically executed on start-up?
-    auto_run = Bool(False)
+        return dirs + files
 
-    #-- Event Handlers -------------------------------------------------------
+    #-- Factory Methods for Creating Tutorial Files Based on File Type --------
 
-    def _path_changed(self, path):
-        """ Creates the appropriate section based on the value of the path.
+    def _parse_avi_file(self, path, name):
```
Comment (corranwebster) [inline traitsui/extras/tutor.py]: This can be safely removed.

File: traitsui/extras/tutor.py [hunk 3]
```diff
@@ -873,729 +191,285 @@ def run_code(self):
...
-    def _add_swf_item(self, path):
-        """ Creates a description item for a Flash file.
-        """
-        if is_windows and wx_available:
-            self.descriptions.append(FlashItem(path=path))
+        return self._parse_jpg_file(path, name)
 
-    def _add_mov_item(self, path):
-        """ Creates a description item for a QuickTime movie file.
+    def _parse_mov_file(self, path, name):
```
Comment (corranwebster) [inline traitsui/extras/tutor.py]: This can be removed

File: traitsui/extras/tutor.py [hunk 4]
```diff
@@ -873,729 +191,285 @@ def run_code(self):
...
-    def _add_avi_item(self, path):
-        """ Creates a description item for an AVI movie file.
-        """
-        self._add_wmv_item(path)
+        html = QTMovieTemplate % (path2, path2)
+        return TutorialHTMLFile(description=html, name=name, parent=self)
 
-    def _add_jpg_item(self, path):
-        """ Creates a description item for a JPEG image file.
+    def _parse_mp3_file(self, path, name):
```
Comment (corranwebster) [inline traitsui/extras/tutor.py]: This can be removed.

File: traitsui/extras/tutor.py [hunk 5]
```diff
@@ -873,729 +191,285 @@ def run_code(self):
...
+        path = self.path
+        for name in os.listdir(path):
+            cur_path = os.path.join(path, name)
+            if os.path.isdir(cur_path):
+                dirs.append(TutorialPath(name=name, parent=self))
+            elif self.use_files:
+                name, ext = os.path.splitext(name)
+
+                # If we have a handler for the file type, invoke it:
+                method = getattr(
```
Comment (corranwebster) [inline traitsui/extras/tutor.py]: If you wanted to do a bit better of a refactor, replacing "magic method names" with a dictionary mapping file extensions to factory functions would likely make the code much cleaner, and might allow something like having a tutorial be the same base class as a demo but with a slightly different set of registered file handlers, assuming you can't get the codebases completely aligned.

Comment (corranwebster) [reply traitsui/extras/tutor.py]: The factory functions could be class methods on the appropriate classes being created.

File: traitsui/extras/tutor.py [hunk 6]
```diff
@@ -873,729 +191,285 @@ def run_code(self):
...
-
-        # Add the resulting item to the descriptions list:
-        self.descriptions.append(item)
+        text = _read_file(path)
+        return TutorialTxtFile(description=text, name=name, parent=self)
 
-    def _create_html_item(self, **traits):
-        """ Creates a platform specific html item and adds it to the list of
-            descriptions.
+    def _parse_wmv_file(self, path, name):
```
Comment (corranwebster) [inline traitsui/extras/tutor.py]: This can safely be removed.

**Comment by shoeb-github** [comment]
> Rather than subclassing, how much of this functionality can be merged into the `demo.py` classes? In other words, can we enhance the demos with tutor functionality without breaking them?

Or alternatively, how much functionality can we drop from the tutor.py code?

Also, I don't think we have _any_ actual tutorials using videos or similar, so we can drop support for a lot of the file formats without losing anything. 



I think it is possible to add this functionality to the demo without altering the demo functionality.

**Comment by shoeb-github** [comment]
closing in favor of #712`

**Pull request closed without merging**